### ISSUE TITLE
Infinite lock issue with RedisLockExtension

### ISSUE THREAD

**Issue by mgoodfellow** [TurnerSoftware/CacheTower#181]
### What does the bug do?

Produces an infinite lock on resource

### How can it be reproduced?

Aha, I'm struggling to build a reproduction, however it can be described via theory.

### Setup

CacheStack 
Redis layer only
RedisLockExtension used

### What happened?

We started using background refresh in production, and we found we were losing MySQL connections - they effectively got taken from the pool and never returned. This process happens slowly - a couple every few 100k requests.

Initially we turned off background refreshing, but issue was still present. We then just used Get/Set functions (as we did previously) and the issue was not there. Finally we removed the `RedisLockExtension` without background refreshing, and the issue was not there, and then added background refreshing back in, and issue still not there. This effectively points the finger at `RedisLockExtension`. We have been stable in production for 48 hours without it (no lost connections from the pool), and prior to this we were losing connections every hour or 2.

Whats going on?

When we build a context via our IoC container, we inject a DB session - under the hood this is grabbing a MySQL connection. Also when most web requests run, we also grab a DB session (and therefore MySQL connection) for the request.

Infinite lock issue

It's taken me a while to spot:

1) Try to get a lock - https://github.com/TurnerSoftware/CacheTower/blob/main/src/CacheTower.Extensions.Redis/RedisLockExtension.cs#L73

2) If we can't get a lock, we wait on the completion source: https://github.com/TurnerSoftware/CacheTower/blob/main/src/CacheTower.Extensions.Redis/RedisLockExtension.cs#L101

3) Once the owner of the lock has refreshed a value, it publishes that via redis: https://github.com/TurnerSoftware/CacheTower/blob/main/src/CacheTower.Extensions.Redis/RedisLockExtension.cs#L80

4) The subs try to clear down that completion source by setting the result: https://github.com/TurnerSoftware/CacheTower/blob/main/src/CacheTower.Extensions.Redis/RedisLockExtension.cs#L50

But what happens if there is a disruption to the Redis connection, and the sub never receives the lock release? Well, in this case you have an infinite lock, and there is no way to release it.

Although a similar process is used in the main CacheStack - https://github.com/TurnerSoftware/CacheTower/blob/main/src/CacheTower/CacheStack.cs#L313 - We aren't relying on an external dependency to transmit a message to make sure we unlock, instead it happens in the thread context that performed the update and this issue cannot occur.

Thoughts:

* We could use a timeout on the completion source? Make sure the lock always releases even if no sub received.
* I wonder if a busy wait approach on LockTakeAsync/LockReleaseAsync would be better, rather than relying on pub/sub?

**Comment by Turnerj** [comment]
I wonder if this relates to the intermittent issues with Redis during CI builds too...

(Will give a more complete reply soon)

**Comment by mgoodfellow** [comment]
> I wonder if this relates to the intermittent issues with Redis during CI builds too...

I was thinking that, although that happens fairly regularly, whereas this happens (in context) extremely rarely. Of course, exact figures are hard because even with high traffic I might not have a high lock contention. I would imagine however that the longer the `getter` function takes to evaluate and the shorter the stale/expiry, the more likely you are to run into this as the lock contention will be higher.

Just having a dig on the tests at the moment to see if anything pops out as obvious as an issue.

I also notice there are edge cases which can occur with subscribers not reconnecting properly, although I do not believe this is [the issue here](https://github.com/StackExchange/StackExchange.Redis/issues/1110) - there does definitely need to be a failsafe, even if it is just `LockTimeout` used to complete the completionSource to make sure we never end up in this position. However, I wonder if we can combine a busy wait fail safe with this pub/sub model to balance performance / recovery. 

I think this quickest fix would simply be to use `LockTimeout` to clear down the completionSource.

**Comment by Turnerj** [comment]
I created a minimum example in [this issue](https://github.com/StackExchange/StackExchange.Redis/issues/1827) on SE.Redis to see if the root cause of the flaky test is the pub/sub side of things. It has been suggested to me to look at [Redis streams](https://redis.io/topics/streams-intro) as an alternative to pub/sub but I think it is just a subtle bug rather than a fundamental issue.

You're right though and I think you found one subtle bug - while a lock had a timeout on Redis, it didn't actually propagate back to the locally locked instances the extension was tracking.

**Comment by Turnerj** [comment]
I've just published a new release (0.11.1) with your fix added - let me know how it goes!

**Comment by mgoodfellow** [comment]
I finally had a chance to put it all back in production, and had the spin lock etc. enabled, but still suffering this issue. The crazy thing is that I'm seeing connections which have been "lost" for minutes/hours.

I still point the finger at the `RedisLockExtension`, because if I remove it everything is fine under our production load, and we've never seen a lost connection, but within a few hours of running it with this extension enabled, connections get lost. Obviously this is still anecdotal, and at best an educated guess.

I need to do some more digging and actually check to see if these lock keys are being removed from redis (the fail safe should always work if they are?)

Otherwise, there is something else at foot.

I will try to find some time to dig more over the next few days/weeks!

Everything is rock solid without this extension, utilising background refresh and our SimpleInjector cache activator. It isn't critical for us to have this distributed lock feature (our lock contention is low, and we don't run huge numbers of pods in our farm), however it is a nice to have, so I will continue to investigate to see if I can get a root cause.

**Comment by Turnerj** [comment]
Thanks for letting me know! I'll try to setup a test of some sort locally to try and recreate the issue.


Logically the code looks correct but it might just be a useful exception is being silently swallowed.

**Comment by Turnerj** [comment]
I know this is way after the fact but I'm curious about one detail with your setup: How many instances of your application were running?

I may have found _another_ cause of infinite locks (well, 1 minute by default) but that should only happen if two requests within the same instance of a `CacheStack` managed to get through to the `RedisLockExtension` internally locking mechanism.

In the `CacheStack`, we can see that a local lock is freed via `UnlockWaitingTasks` when we get a value or when we crash: https://github.com/TurnerSoftware/CacheTower/blob/24e3022b85e71c6db6d194fe4baa4b398d41db86/src/CacheTower/CacheStack.cs#L308-L333

In the `RedisLockExtension`, we don't free the local locks we establish within the `RedisLockExtension`: https://github.com/TurnerSoftware/CacheTower/blob/24e3022b85e71c6db6d194fe4baa4b398d41db86/src/CacheTower.Extensions.Redis/RedisLockExtension.cs#L78-L87

This means when the path that has the lock actually gets the new data to cache, it never lets any other code that is locally locked within the same `RedisLockExtension` instance to be free. It _shouldn't_ have both paths be hit on the same cache key for a `CacheStack` at the same time but maybe there is another race condition I've missed.

**Comment by mgoodfellow** [comment]
Hey @Turnerj ,

Interesting - in this particular API we run 4 instances in a K8s cluster, but we are using this cache on a hot path of requests (10s per second), with a relatively expensive cache to build.

We happily use this currently with redis, just without this lock extension - which means that we can have the expense of multiple instances building the cache simultaneously, but that tradeoff vs thread starvation is preferable. In fact, we don't really notice a performance hit when we sometimes build the cache multiple times on different instances. 

Of course, it would always be great to _know_ the answer, so it would be interesting to patch this, and I'm happy to spin it into production to see if it fixes it.
